src/Interpreters/QueryMetricLog.cpp

```diff
@@ -139,51 +199,66 @@ void QueryMetricLog::finishQuery(const String & query_id, TimePoint finish_time,
-        lock.unlock();
+        global_lock.unlock();
     }
 }
 
-std::optional<QueryMetricLogElement> QueryMetricLog::createLogMetricElement(const String & query_id, const QueryStatusInfo & query_info, TimePoint query_info_time, bool schedule_next)
+void QueryMetricLogStatus::scheduleNext(String query_id)
 {
-    /// fmtlib supports subsecond formatting in 10.0.0. We're in 9.1.0, so we need to add the milliseconds ourselves.
-    auto seconds = std::chrono::time_point_cast<std::chrono::seconds>(query_info_time);
-    auto microseconds = std::chrono::duration_cast<std::chrono::microseconds>(query_info_time - seconds).count();
-    LOG_DEBUG(logger, "Collecting query_metric_log for query {} with QueryStatusInfo from {:%Y.%m.%d %H:%M:%S}.{:06}. Schedule next: {}", query_id, seconds, microseconds, schedule_next);
-
-    std::unique_lock lock(queries_mutex);
-    auto query_status_it = queries.find(query_id);
-
-    /// The query might have finished while the scheduled task is running.
-    if (query_status_it == queries.end())
+    next_collect_time += std::chrono::milliseconds(interval_milliseconds);
+    const auto now = std::chrono::system_clock::now();
+    if (next_collect_time > now)
     {
-        lock.unlock();
-        LOG_TRACE(logger, "Query {} finished already while this collecting task was running", query_id);
-        return {};
+        const auto wait_time = std::chrono::duration_cast<std::chrono::milliseconds>(next_collect_time - now).count();
+        task->scheduleAfter(wait_time);
+    }
+    else
+    {
+        LOG_TRACE(logger, "The next collecting task for query {} should have already run at {}. Scheduling it right now",
+            query_id, timePointToString(next_collect_time));
+        task->schedule();
     }
+}
+
+std::optional<QueryMetricLogElement> QueryMetricLogStatus::createLogMetricElement(const String & query_id, const QueryStatusInfo & query_info, TimePoint query_info_time, bool schedule_next)
+{
+    /// TODO: remove trace before 24.11 release after checking everything is fine on the CI
+    LOG_TRACE(logger, "Collecting query_metric_log for query {} and interval {} ms with QueryStatusInfo from {}. Next collection time: {}",
+        query_id, interval_milliseconds, timePointToString(query_info_time),
+        schedule_next ? timePointToString(next_collect_time + std::chrono::milliseconds(interval_milliseconds)) : "finished");
 
-    auto & query_status = query_status_it->second;
-    if (query_info_time <= query_status.last_collect_time)
+    if (query_info_time <= last_collect_time)
     {
-        lock.unlock();
+        /// TODO: remove trace before 24.11 release after checking everything is fine on the CI
         LOG_TRACE(logger, "Query {} has a more recent metrics collected. Skipping this one", query_id);
         return {};
     }
 
-    query_status.last_collect_time = query_info_time;
+    /// Leave some margin because task->scheduleAfter takes a value in milliseconds.
```

Comment:
i don't think it will work in real world, actual wake up time might be postponed by a lot because of high load or just scheduling

Reply:
>actual wake up time might be postponed by a lot because of high load or just scheduling

I completely agree, but take into account that this checks exactly the opposite: that the task is never scheduled **before**  it's supposed to. Which is something that should never happen according to [this](https://github.com/ClickHouse/ClickHouse/blob/dc92242c39504bd21f756939854ae7d7ffc41252/src/Core/BackgroundSchedulePool.cpp#L340).

I did see some unexpected behavior, so I decided to add this just to check. I can add more margin just in case or even remove it if you deem so, but I think it `BackgroundSchedulePool` is doing its job, it should never get here.

Reply:
> that the task is never scheduled before it's supposed to

sorry, i misunderstood it